### PR TITLE
Joe Jiang

### PR DESCRIPTION
**Pull request by ninesky00**

### CODE REVIEW

**Comment by KathleenScriver** [comment]
Moving forward, let's make the commit messages just describe the functionality they're adding, not the iterations. We'll also want to start doing a commit for each test that's written, then a commit once the code is written that makes that test pass.

File: lib/library.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  def initialize(name)
+    @name = name
+    @books = []
+    @authors = []    
+    @checked_out_books = []
+    @checkout_log = []
+  end
+
+  def add_author(author)
+    #an add_book method is probably more proper as a library is unlikely to add all books of an author at once.
```
Comment (KathleenScriver) [inline lib/library.rb:12]: Also, what happens when the author writes a new book after they're added to the library? We'd need some way to account for that. If we act under the assumption that a library will hold every book from an author, then we can just calculate those books based off of the @authors array.

File: lib/library.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    author.books.each { |book| @books << book }
+    @authors << author
+  end
+
+  def publication_time_frame_for(author)
+    time_frame = {}
+    time_frame[:start] = author.books.map {|book| book.publication_year}.min
+    time_frame[:end] = author.books.map {|book| book.publication_year}.max
+    time_frame
```
Comment (KathleenScriver) [inline lib/library.rb:21]: Great job with this, however you have repeated code. You're mapping over author.books twice. Can dry that up by storing it in a variable and then calling .min and .max on that variable.

File: lib/library.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    if @checked_out_books.include?(book)
+      @books << book
+      @checked_out_books.delete(book)
+    end
+    #did not include an else statement for the sake of inappropriate checkouts or lost/stolen and not on record
+  end
+
+  def most_popular_book
+    checkout_count = {}
+    @checkout_log.each { |book| checkout_count[book] ? checkout_count[book] += 1 : checkout_count[book] = 1 }
```
Comment (KathleenScriver) [inline lib/library.rb:45]: Let's try to avoid turinary statements in Ruby. They're not too readable.

File: test/library_test.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    assert_equal [@charlotte_bronte, @harper_lee], @dpl.authors
+    expected = [@jane_eyre, @professor, @villette, @mockingbird]
+    assert_equal expected, @dpl.books
+    expected = {:start=>"1847", :end=>"1857"}
+    assert_equal expected, @dpl.publication_time_frame_for(@charlotte_bronte)
+    expected = {:start=>"1960", :end=>"1960"}
+    assert_equal expected, @dpl.publication_time_frame_for(@harper_lee)
+  end
+
+  def test_can_only_check_out_books_if_they_exist_and_are_not_checked_out
```
Comment (KathleenScriver) [inline test/library_test.rb:38]: For testing, remember each method should have it's own test. Even if the methods work together, we want to test them one at a time.

**Review by KathleenScriver — COMMENTED** [review]
overall, great job! Your score is a 4.

File: lib/library.rb (hunk 2 above)
Comment (ninesky00) [reply lib/library.rb:21]: omg, was not thinking of scalability here. Thank you for the advice!

File: lib/library.rb (hunk 3 above)
Comment (ninesky00) [reply lib/library.rb:45]: okay :( , but it looks super cool. :(

File: test/library_test.rb (hunk 4 above)
Comment (ninesky00) [reply test/library_test.rb:38]: will make sure to try to do that going forward. Thanks

**Pull request closed without merging**